Added `iter_prop` and `iter_geometry` to Geojsonm

IIITM-Jay opened the pull request:
## Description
This is in continuation with #7. This PR adds up the `iter_geometry` and `iter_prop` to Geojsonm.

## Changes Incorporated

- [x] Added `iter_prop` and `iter_geometry` functions to `geojsonm.ml` file
- [x] Made respective changes to the `geojsonm.mli` file

IIITM-Jay commented:
Hi Sir @patricoferris , May I request you to kindly look into this . Can you please point me out on what I am missing

patricoferris commented on src/geojsonm/geojsonm.mli:
```diff
@@ -79,3 +79,6 @@ val fold_geometry :
   ('a -> Ezjsonm.value -> 'a) -> 'a -> Jsonm.src -> ('a, Err.t) result
 (** [fold_props f init src] *)
+
+val iter_geometry : (G.Geometry.t -> unit) -> Jsonm.src -> (unit, Err.t) result
```
Some documentation in the same style as `map_*` and `fold_*`  would be great here.

IIITM-Jay replied:
Yes Sir, @patricoferris done now,

patricoferris commented on src/geojsonm/geojsonm.ml:
```diff
@@ -245,3 +245,42 @@ let fold_props f init src =
+        | Error (`Msg m) -> raise (Abort (`Unexpected m))
+        | Ok g ->
+            let g' = f g in
+            enc (`Lexeme t);
```
Iterating doesn't produce a new GeoJSON object (unlike mapping) so we don't need to encode the data (the function has also been removed hence the CI errors).

IIITM-Jay replied:
Hi @patricoferris Thank you sir for letting me know this!

patricoferris commented on src/geojsonm/geojsonm.ml:
```diff
@@ -245,3 +245,42 @@ let fold_props f init src =
+        match G.Geometry.of_json @@ decode_single_object decoder with
+        | Error (`Msg m) -> raise (Abort (`Unexpected m))
+        | Ok g ->
+            let g' = f g in
```
`f g` will be of type unit if I'm not mistaken, so `f g;` is sufficient (no need to bind it to a variable).

IIITM-Jay replied:
Yeah my mistake , Thank you so much for pointing me this! :pray:

patricoferris reviewed (changes requested):
Thanks @IIITM-Jay -- some comments, I think there might still be a little confusion between mapping and iterating. I think this tutorial does a good job of explaining it in the context of Lists (the idea generalises): https://v3.ocaml.org/docs/lists

IIITM-Jay commented on CONTRIBUTING.md:
````diff
@@ -23,7 +23,22 @@ And to run the tests (which are in the `test` directory) run:
 ```
 
-If you hit any problems please feel free to open an issue. Please also run the formatter on the code too with `dune build @fmt --auto`, you might need to install the correct version of `ocamlformat`.
+
````
Yeah Done Sir, Thank you so much ! :pray:

IIITM-Jay commented:
Hi @patricoferris Sir, Thank you so much for giving your valuable time for reviewing this and also for giving me some knowledge based on `iter` and `map`.

I have incorporated those changes now. May I request you to have a look Sir.

Thank you so much once again! :pray:

patricoferris commented on src/geojsonm/geojsonm.mli:
```diff
@@ -79,3 +79,16 @@ val fold_geometry :
+
+(** {2 Iterators}
+
+    Iterators works much more similar way as that of Maps. It takes an
+    imperative function of type 'a -> unit and an 'a list and applies the
+    function to every element in turn. The only difference is that Iterators
+    doesn't produce a new GeoJSON object (unlike mapping). *)
```
This is copied from Lists I think so doesn't make a lot of sense in this context :)) 

```suggestion
     Iterators are similar to map functions except they a function [f] that takes a single
     element from the data-structure we are iterating over that must return [unit]. In that 
     sense they tend to functions with side-effects, such as [print_endline].
```

Something like that might be better.

patricoferris commented on src/geojsonm/geojsonm.ml:
```diff
@@ -245,3 +245,37 @@ let fold_props f init src =
+  let rec go () =
+    match Jsonm.decode decoder with
+    | `Lexeme (`Name "properties") ->
+        let o = f @@ decode_single_object decoder in
+        o;
```
```suggestion
        f @@ decode_single_object decoder;
```

patricoferris reviewed (changes requested):
Thanks! This is looking better, some more changes. Could you also maybe add a test too?

IIITM-Jay commented:
Hi Sir, @patricoferris , oh by mistakenly you forgot to close the comment that's why showing error, but no issues I will fix them and try to add some more on it.

IIITM-Jay commented:
Hi Sir @patricoferris , I have incorporated your suggestive changes and added some few lines _(if you would like it)_. May I request you to kindly have a look Sir.

Moreover, what I am thinking is once this PR gets merged, it would be better to update the [documentation for Goejsonm Module](https://geocaml.github.io/ocaml-geojson/geojsonm/Geojsonm/index.html) which for the time being now, only includes `maps` and `folds`. It would be better to add `iters` too on that page. Probably I will open up a new issue, if you permit and allow me to do so, and for that as I must guess we have to work on this on some different branch, if I'm not wrong.

Regarding adding of the test cases, is it okay if I will do this later after having done with the above tasks, as I have already completed much before writing of the tests for two geometry `Point` and `LineString` so as to get the sequence as (`Point`, `MultiPoint`, `LineString`, `MultiLineString`) , but I haven't committed as I thought this will be considered as smaller contribution ,out of which two of them are already present for now. so that others can take the rest other geometries. Having said that I don't have any problem if new contributors have to work on that issue. I am thinking to add tests for `map`, `fold` & `iter` into a fresh new PR later on. 

**_To summarise_**,  the workflow that I am thinking is : 

**Merging of this PR**   -->   **Adding documentation on the `Geojsonm` Module web page**   -->   **working on test cases**

Sir, I would really like hear your thoughts on the above points.

patricoferris commented on src/geojsonm/geojsonm.mli:
```diff
@@ -79,3 +79,20 @@ val fold_geometry :
+
+    A standard usage of 'Iterators' is as follows:
+
+    val iter : ('a -> unit) -> 'a list -> unit
```
Could we have a concrete example here instead of a type signature. A type isn't really a standard usage but something like:

```ocaml
List.iter print_int [1; 2; 3; 4]
```

patricoferris replied:
This example was meant to be more of an inspiration that to be used in the documentation, sorry if that was confusing. We're not iterating over a list but instead over `geometry` objects and `properties`. 

For example the code you wrote in the the test would be better so long as it compiles (see the other comment).

Also please remove any mention of `val iter : ('a -> unit) -> 'a list -> unit`, we are documenting the type so it should be clear.

patricoferris commented on src/geojsonm/geojsonm.mli:
```diff
@@ -79,3 +79,20 @@ val fold_geometry :
+
+    A standard usage of 'Iterators' is as follows:
+
+    val iter : ('a -> unit) -> 'a list -> unit
```
Could we have a concrete example here instead of a type signature. A type isn't really a standard usage but something like:

```ocaml
List.iter print_int [1; 2; 3; 4]
```

patricoferris commented:
Thanks for the updates :))

> Moreover, what I am thinking is once this PR gets merged, it would be better to update the [documentation for Goejsonm Module](https://geocaml.github.io/ocaml-geojson/geojsonm/Geojsonm/index.html) which for the time being now, only includes maps and folds. It would be better to add iters too on that page.

Don't worry about that, I can update the documentation at some point in the future, it's a pretty automated process :)) Once we get a release done, it'll also be available on v3.ocaml.org too.

> Regarding adding of the test cases for above, is it okay if I will do this later after having this PR merged...

I don't quite see why it shouldn't be a part of this PR? I'd rather merge new code with confidence it works and not splitting up things that are part of the same feature into multiple PRs if that's okay?

> If you don't mind, I am thinking to add tests for map, fold & iter into a fresh new PR later on...

`map` and `fold` are already somewhat tested (see https://github.com/geocaml/ocaml-geojson/blob/main/test/geojsonm/expect/test.ml) so I think it would be good as part of this PR to add an iter test. Note it doesn't have to match the `map` and `fold` ones. It can be in a separate file if you like :))

IIITM-Jay commented:
> 
> I don't quite see why it shouldn't be a part of this PR? I'd rather merge new code with confidence it works and not splitting up things that are part of the same feature into multiple PRs if that's okay?
> 
Hi Sir @patricoferris , Good Morning, truely said sir. It's always a better practice to merge the new features along with its verification through test cases. I will soon push the commit for the same.
> 
> `map` and `fold` are already somewhat tested (see https://github.com/geocaml/ocaml-geojson/blob/main/test/geojsonm/expect/test.ml) so I think it would be good as part of this PR to add an iter test. Note it doesn't have to match the `map` and `fold` ones. It can be in a separate file if you like :))
Sure Sir, I will follow this.

Thank you once again Sir, for your time Sir despite of your busy schedules.:pray:

IIITM-Jay commented:
Hi Sir @patricoferris , I have added the example that you have mentioned and also added the test in a new file. _(in the commit [Added more concrete example and test case for iters](https://github.com/geocaml/ocaml-geojson/pull/25/commits/bc54c442b94f466e09125f7efc11af2163f835ba) )_

May I request you to kindly have a look in those and point me out if I am missing something :pray: 
Waiting for your feedback and suggestive changes into that

Thanking You.

patricoferris commented on src/geojsonm/geojsonm.mli:
```diff
@@ -79,3 +79,26 @@ val fold_geometry :
+    Iterators are similar to map functions except they take a function [f] that
+    takes a single element from the data-structure as an argument and returns
+    [unit]. In that sense, they tend to be functions with side-effects, such as
+    [print_endline], [print_newline ()] *)
```
```suggestion
    [print_endline] *)
```

patricoferris commented on src/geojsonm/geojsonm.mli:
```diff
@@ -79,3 +79,26 @@ val fold_geometry :
+(** [iter_geometry f src] will apply [f] to all GeoJson objects. *)
+
+val iter_props : (Ezjsonm.value -> unit) -> Jsonm.src -> (unit, Err.t) result
+(** [iter_props src ~f] will apply [f] to each feature's properties field. *)
```
In the small code example (`[iter_props src ~f]`) you've used a label for the function but the type signature shows no labels for this function.

patricoferris commented on test/geojsonm/expect/testIters.ml:
```diff
@@ -0,0 +1,39 @@
+let capitalise_nom obj =
```
Can you name the file `test_iters` and also add a [test stanza](https://dune.readthedocs.io/en/stable/dune-files.html#test) for this code (at the moment this code is not being compiled or run I think).

patricoferris replied:
Please also delete this file :))

IIITM-Jay commented:
Thank you Sir @patricoferris , for pointing me out to add stanza , (_Just skipped from mind previously and assumed that it was compiling and running well),_

May I request you to kindly help me out by letting me know what I'm missing as why it's not passing ...

Many many thanks to you for taking your time for reviewing and apologizing also for disturbing you on weekends :pray:

IIITM-Jay commented:
Hi @patricoferris sir, May I request you to please help me out with this. :pray:

patricoferris commented on test/geojsonm/expect/test_iters.ml:
```diff
@@ -0,0 +1,41 @@
+  print_or_fail
+    ( with_src "./input/simple.geojson" @@ fun src ->
+      Geojsonm.iter_geometry remove_coords src );
+  `Buffer dst;
```
```suggestion
```

IIITM-Jay replied:
I have tried this before, but got some issues like , **_unused struct variable t_**; like that
Hi Sir @patricoferris , Ihave just left a comment when trying some other way.... can you please look into it once.

patricoferris commented on test/geojsonm/expect/test_iters.ml:
```diff
@@ -0,0 +1,41 @@
+  Buffer.clear dst;
+  print_or_fail
+    ( with_src "./input/simple.geojson" @@ fun src ->
+      Geojsonm.iter_props capitalise_nom src (`Buffer dst) );
```
The type of the `iter` functions have two arguments. The source to read the GeoJSON from and the function to apply to each element depending on what we are iterating on. Here you have also given it a destination to write things to (`` `Buffer dst``) which is no longer needed.

patricoferris replied:
Additionally, I think you will run into problems because the `capitalise_nom` function does not return `unit`.

IIITM-Jay replied:
Yeah, that's true @patricoferris Sir, but how can we convert the type into unit.... it's a bit getting confusing for me...

patricoferris reviewed (commented):
Thanks @IIITM-Jay :))

The tests have retained too much of the logic from mapping which is what is causing the issues, if we can just get them to compile I'm happy to merge and I can maybe expand them a little bit later to do more testing :))

IIITM-Jay commented:
Hi @patricoferris Sir, WhenI am trying the following :
```
let remove_coords p =
  let open Geojsonm in
  match p with G.Geometry.Polygon _ -> G.Geometry.(()) | p -> ()

let with_src f func =
  let ic = open_in f in
  let res =
    try func (`Channel ic)
    with e ->
      close_in ic;
      raise e
  in
  close_in ic;
  res

let () =
  let print_or_fail = function
    | Ok () -> Format.printf "%s\n\n"
    | Error e ->
        Geojsonm.Err.pp Format.err_formatter e;
        failwith "Internal err"
  in
  print_or_fail
    (( with_src "./input/simple.geojson" @@ fun src ->
       Geojsonm.iter_geometry remove_coords );
     src)
```

I am getting the following error:
![image](https://user-images.githubusercontent.com/65283880/162630653-eff1d35d-f84e-4835-a51a-6d28c29341bc.png)

IIITM-Jay commented:
Hi @patricoferris Sir, When I am trying the following:
```
let remove_coords p =
  let open Geojsonm in
  match p with G.Geometry.Polygon _ -> G.Geometry.(()) | p -> ()

let with_src f func =
  let ic = open_in f in
  let res =
    try func (`Channel ic)
    with e ->
      close_in ic;
      raise e
  in
  close_in ic;
  res

let () =
  let print_or_fail = function
    | Ok () -> Format.printf "%s\n\n"
    | Error e ->
        Geojsonm.Err.pp Format.err_formatter e;
        failwith "Internal err"
  in
  print_or_fail
    (( with_src "./input/simple.geojson" @@ fun src ->
       Geojsonm.iter_geometry remove_coords );
     src)
```

I am getting the following error:

![image](https://user-images.githubusercontent.com/65283880/162630774-98780c23-ed66-4fef-a11f-78b76de727e9.png)

Can you please help me out with the same

patricoferris commented:
Hi @IIITM-Jay

I think my explanations aren't great, apologies. To help speed things along a little I've implemented some very simple tests in this commit: https://github.com/geocaml/ocaml-geojson/commit/774fa44dc5a171ad93665ac3a766dc4a5d0b1b6a

I'm hoping that can help shed a bit of light on the kind of thing I was talking about? Feel free to cherry-pick that commit if it helps at all :))

IIITM-Jay commented:
Hi Sir @patricoferris , Thank you so much for giving your valuable time in this. _I have made the suggestive changes and also **added a bit documentation** (hope you will like it :slightly_smiling_face: :blush: )_....want your feedback in that... :))


> I'm hoping that can help shed a bit of light on the kind of thing I was talking about? Feel free to cherry-pick that commit if it helps at all :))

**Once again Thank you so much :pray:  for clarification on the same and now I also got your mentioning and the use of `print_endline` in this context.**

IIITM-Jay commented:
Hi Sir @patricoferris , Thank you so much for giving your valuable time in this. _I have made the suggestive changes and also **added a bit documentation** (hope you will like it :slightly_smiling_face: :blush: )_....want your feedback in that... :))


> I'm hoping that can help shed a bit of light on the kind of thing I was talking about? Feel free to cherry-pick that commit if it helps at all :))

_Once again Thank you so much :pray:  for clarification on the same , that helped me a lot ...and now I also got your mentioning and the use of **`print_endline`** in this context._

patricoferris reviewed (approved):
Thanks! I think I'll make a few changes in a follow-up PR but otherwise this is great :))
